pkg/routeagent_driver/handlers/ovn/avoid_snat_handler.go

```diff
@@ -0,0 +1,211 @@
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	enableSNATHandlerKey = "enable-snat-handler"
+	remoteCIDRIPSetIPv4  = "SUBMARINER-REMOTECIDRS"
+	localCIDRIPSetIPv4   = "SUBMARINER-LOCALCIDRS"
+)
+
+type AvoidSNATHandler struct {
+	event.HandlerBase
+	pFilter     packetfilter.Interface
+	remoteIPSet packetfilter.NamedSet
+	localIPSet  packetfilter.NamedSet
+}
+
+func NeedToEnableAvoidSNATHandler(cm *corev1.ConfigMap) (bool, error) {
+	enableHandler := false
+
+	if cm != nil {
+		if value, ok := cm.Data[enableSNATHandlerKey]; ok {
+			var err error
+
+			enableHandler, err = strconv.ParseBool(value)
+			if err != nil {
+				return false, errors.Wrapf(err, "unable to parse %q from ConfigMap %q", enableSNATHandlerKey, cm.Name)
+			}
+		}
+	}
+
+	return enableHandler, nil
+}
+
+func NewAvoidSNATHandler() event.Handler {
+
+	h := &AvoidSNATHandler{}
+
+	return h
+}
+
+func (h *AvoidSNATHandler) Init(_ context.Context) error {
+	logger.Info("Starting AvoidSNATHandler")
+
+	var err error
+
+	h.pFilter, err = packetfilter.NewWithDriver(nftables.New)
+	if err != nil {
+		return errors.Wrap(err, "error initializing packetfilter with nftables driver")
+	}
+
+	if err := h.pFilter.CreateIPHookChainIfNotExists(&packetfilter.ChainIPHook{
+		Name:     constants.SmSelfSnatChain,
+		Type:     packetfilter.ChainTypeNAT,
+		Hook:     packetfilter.ChainHookPostrouting,
+		Priority: packetfilter.ChainPriorityMiddle,
+	}); err != nil {
+		return errors.Wrapf(err, "error creating IPHookChain chain %s", constants.SmSelfSnatChain)
+	}
+
+	return h.createSelfSNATRules()
+}
+
+func (h *AvoidSNATHandler) GetName() string {
+	return "submariner-avoid-snat-handler"
+}
+
+func (h *AvoidSNATHandler) GetNetworkPlugins() []string {
+	return []string{cni.OVNKubernetes}
+}
+
+func (h *AvoidSNATHandler) newNamedSetSet(key string) packetfilter.NamedSet {
+	return h.pFilter.NewNamedSet(&packetfilter.SetInfo{
+		Name: key,
+	})
+}
+
+func (h *AvoidSNATHandler) createSelfSNATRules() error {
+	h.remoteIPSet = h.newNamedSetSet(remoteCIDRIPSetIPv4)
+	if err := h.remoteIPSet.Create(true); err != nil {
+		return errors.Wrapf(err, "error creating set %q", remoteCIDRIPSetIPv4)
+	}
+
+	h.localIPSet = h.newNamedSetSet(localCIDRIPSetIPv4)
+	if err := h.localIPSet.Create(true); err != nil {
+		return errors.Wrapf(err, "error creating set %q", localCIDRIPSetIPv4)
+	}
+
+	logger.Info("Creating packetfilter self-snat rules")
+
+	ruleSpecIngress := packetfilter.Rule{
+		SrcSetName:  remoteCIDRIPSetIPv4,
+		DestSetName: localCIDRIPSetIPv4,
+		Action:      packetfilter.RuleActionSelfSNAT,
+	}
+
+	if err := h.pFilter.AppendUnique(packetfilter.TableTypeNAT, constants.SmSelfSnatChain, &ruleSpecIngress); err != nil {
+		return errors.Wrapf(err, "unable to append rule %+v", &ruleSpecIngress)
+	}
+
+	return nil
+}
+
+func (h *AvoidSNATHandler) LocalEndpointCreated(endpoint *submV1.Endpoint) error {
+	subnets := cidr.ExtractIPv4Subnets(endpoint.Spec.Subnets)
```

Comment:
Just curious: Are we not supporting dualStack clusters as the code is only extracting IPv6 subnets.